src/Template.php

```diff
@@ -371,36 +313,46 @@ public function getBlocks()
         return $this->blocks;
     }
 
-    public function display(array $context, array $blocks = [])
+    public function display(array $context, array $blocks = []): void
     {
-        $this->displayWithErrorHandling($this->env->mergeGlobals($context), array_merge($this->blocks, $blocks));
+        foreach ($this->yield($context, $blocks) as $data) {
+            echo $data;
+        }
     }
 
-    public function render(array $context)
+    public function render(array $context): string
     {
-        $level = ob_get_level();
-        if ($this->env->isDebug()) {
-            ob_start();
-        } else {
-            ob_start(function () { return ''; });
+        $content = '';
+        foreach ($this->yield($context) as $data) {
+            $content .= $data;
         }
-        try {
-            $this->display($context);
-        } catch (\Throwable $e) {
-            while (ob_get_level() > $level) {
-                ob_end_clean();
-            }
 
-            throw $e;
-        }
-
-        return ob_get_clean();
+        return $content;
     }
 
-    protected function displayWithErrorHandling(array $context, array $blocks = [])
+    /**
+     * @return iterable<string>
+     */
+    public function yield(array $context, array $blocks = []): iterable
     {
+        $context = $this->env->mergeGlobals($context);
+        $blocks = array_merge($this->blocks, $blocks);
+        $level = ob_get_level();
+        ob_start();
+
         try {
-            $this->doDisplay($context, $blocks);
+            foreach ($this->doDisplay($context, $blocks) as $data) {
+                if (ob_get_length()) {
+                    $data = ob_get_clean().$data;
```

Comment:
This is the meat of the PR

Reply:
Please add tests covering this behavior though. I don't think there is any right now.

Reply:
Test added